types/src/on_chain_config.rs

```diff
@@ -9,7 +9,38 @@ use crate::{
 };
 use libra_crypto::HashValue;
 use move_core_types::identifier::Identifier;
-use serde::{Deserialize, Serialize};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+
+/// Trait to be implemented by a storage type from which to read on-chain configs
+#[async_trait::async_trait(?Send)]
+pub trait ConfigStorage {
+    async fn fetch_config(&self, access_path: AccessPath) -> Option<Vec<u8>>;
+}
+
+/// Trait to be implemented by a Rust struct representation of an on-chain config
+/// that is stored in storage as a deserialized byte array
+#[async_trait::async_trait(?Send)]
+pub trait OnChainConfig {
+    const IDENTIFIER: &'static str;
+
+    async fn fetch_config<U: ConfigStorage>(storage: &U) -> anyhow::Result<Self>
+    where
+        Self: DeserializeOwned,
+    {
+        let access_path = access_path_for_config(Identifier::new(Self::IDENTIFIER).unwrap());
+        let bytes = storage
+            .fetch_config(access_path)
+            .await
+            .ok_or_else(|| anyhow::format_err!("no config found"))?;
+        let result = lcs::from_bytes::<Self>(&bytes).map_err(|e| {
```

Comment:
This won't really work I think. In the example of VMPublishingOption, the move representation will be a plain Vec<u8>, thus we need an extra round of deserialization... see [here](https://github.com/libra/libra/blob/f8d8a9a94480f546a1bee3146f2c9173f66ed1d1/language/libra-vm/src/on_chain_configs.rs#L30). Having this clean trait implementation is indeed appealing tho.

Reply:
We might separate the deserialization logic out as another function in this trait and have it default implemented by lcs. In the VMPublishingOption example we can override that function with a custom deserializer.

Reply:
> This won't really work I think. In the example of VMPublishingOption, the move representation will be a plain Vec, thus we need an extra round of deserialization... see [here](https://github.com/libra/libra/blob/f8d8a9a94480f546a1bee3146f2c9173f66ed1d1/language/libra-vm/src/on_chain_configs.rs#L30). Having this clean trait implementation is indeed appealing tho.

I think I handle the first deserialization in my implementation of `ConfigStorage` for `RemoteCache` in this PR? Could you check if that's right?

Reply:
Ah you are right! I thought that we might not always want to have the top level vector<u8> wrapper. The reason is that having a vector<u8> wrapper will prevent those values to be interpreted on the Move side.